### PR TITLE
[SEDONA-426] Change raster cloning to be able to include metadata.

### PR DESCRIPTION
**Pull request by iGN5117**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-426. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?
- Add metadata preservation in raster cloning

## How was this patch tested?
- Passed current tests. 

## Did this PR include necessary documentation updates?

- Yes, I am adding a new API. I am using the [current SNAPSHOT version number](https://github.com/apache/sedona/blob/99239524f17389fc4ae9548ea88756f8ea538bb9/pom.xml#L29) in since `vX.Y.Z` format.

### CODE REVIEW

File: common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java [hunk 1]
```diff
@@ -80,6 +79,88 @@ public static GridCoverage2D create(WritableRaster raster, GridGeometry2D gridGe
...
+        ColorModel originalColorModel = referenceRaster.getRenderedImage().getColorModel();
+        GridGeometry2D gridGeometry = referenceRaster.getGridGeometry();
+        int numBand = raster.getNumBands();
+        int rasterDataType = raster.getDataBuffer().getDataType();
+
+        final ColorSpace cs = new BogusColorSpace(numBand);
+        final int[] nBits = new int[numBand];
+        Arrays.fill(nBits, DataBuffer.getDataTypeSize(rasterDataType));
+        ColorModel colorModel =
+                new ComponentColorModel(cs, nBits, false, true, Transparency.OPAQUE, rasterDataType);
```
[the remark this answers is not in the document]
Comment (iGN5117) [reply common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java]: Thanks, pushed the change

**Comment by jiayuasu** [comment]
@iGN5117 Do we also need to update other related functions such as `copyRasterAndAppendBand` and `copyRasterAndReplaceBand` since they also clone the rasters?

File: common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java [hunk 2]
```diff
@@ -80,6 +79,89 @@ public static GridCoverage2D create(WritableRaster raster, GridGeometry2D gridGe
...
+            propertyMap = referenceRaster.getProperties();
+        }
+        ColorModel originalColorModel = referenceRaster.getRenderedImage().getColorModel();
+        GridGeometry2D gridGeometry = referenceRaster.getGridGeometry();
+        int numBand = raster.getNumBands();
+        int rasterDataType = raster.getDataBuffer().getDataType();
+
+        final ColorSpace cs = new BogusColorSpace(numBand);
+        final int[] nBits = new int[numBand];
+        Arrays.fill(nBits, DataBuffer.getDataTypeSize(rasterDataType));
```
Comment (Kontinuation) [inline common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java:105]: I believe that these could also be moved to the else branch

Comment (iGN5117) [reply common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java]: Sorry about that! 
Pushed the change

**Comment by iGN5117** [comment]
> @iGN5117 Do we also need to update other related functions such as `copyRasterAndAppendBand` and `copyRasterAndReplaceBand` since they also clone the rasters?

Yes @jiayuasu I will do that once the new rasterUtils functions are approved, This PR will finally replace all places where rasters are cloned

File: spark/common/src/test/scala/org/apache/sedona/sql/rasteralgebraTest.scala [hunk 3]
```diff
@@ -1074,7 +1074,8 @@ class rasteralgebraTest extends TestBaseScala with BeforeAndAfter with GivenWhen
           }
           assert(rast2.getRenderedImage.getSampleModel.getDataType == dataType)
           val noDataValue = RasterUtils.getNoDataValue(rast2.getSampleDimension(0))
-          assert(noDataValue.isNaN)
+          val originalNoDataValue = RasterUtils.getNoDataValue(rast.getSampleDimension(0))
```
Comment (Kontinuation) [inline spark/common/src/test/scala/org/apache/sedona/sql/rasteralgebraTest.scala:1077]: The newly added original no data value is not used.

File: common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java [hunk 4]
```diff
@@ -81,12 +79,100 @@ public static GridCoverage2D create(WritableRaster raster, GridGeometry2D gridGe
...
+                    newBands[k] = createSampleDimensionWithNoDataValue("band_" + k, noDataValue);
+                }
+            }
+            bands = newBands;
+        }
+        GridCoverage2D[] referenceRasterSources = keepMetadata ? referenceRaster.getSources().toArray(new GridCoverage2D[0]) : null;
+        Map propertyMap = null;
+        if (keepMetadata) {
+            propertyMap = referenceRaster.getProperties();
+        }        CharSequence rasterName = keepMetadata ? referenceRaster.getName() : "genericCoverage";
```
Comment (Kontinuation) [inline common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java:165]:
```suggestion
        }
        CharSequence rasterName = keepMetadata ? referenceRaster.getName() : "genericCoverage";
```

File: common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java [hunk 5]
```diff
@@ -81,12 +79,100 @@ public static GridCoverage2D create(WritableRaster raster, GridGeometry2D gridGe
...
+        return gridCoverageFactory.create(rasterName, image, gridGeometry2D, bands, referenceRasterSources, propertyMap);
+    }
+
+        /**
+         * Create a new empty raster from the given WritableRaster object.
+         * @param raster The raster object to be wrapped as an image.
+         * @param gridGeometry The grid geometry of the raster.
+         * @param bands The bands of the raster.
+         * @return A new GridCoverage2D object.
+         */
```
Comment (Kontinuation) [inline common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java:175]:
```suggestion
    /**
     * Create a new empty raster from the given WritableRaster object.
     * @param raster The raster object to be wrapped as an image.
     * @param gridGeometry The grid geometry of the raster.
     * @param bands The bands of the raster.
     * @return A new GridCoverage2D object.
     */
```

File: spark/common/src/test/scala/org/apache/sedona/sql/rasteralgebraTest.scala (hunk 3 above)
Comment (iGN5117) [reply spark/common/src/test/scala/org/apache/sedona/sql/rasteralgebraTest.scala]: Removed, thanks!

File: common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java (hunk 4 above)
Comment (iGN5117) [reply common/src/main/java/org/apache/sedona/common/utils/RasterUtils.java]: Some incorrect formattings creeped up after I renamed, fixed. Thanks!

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**